service/worker/batcher/activities.go

```diff
@@ -0,0 +1,276 @@
+// The MIT License
+//
+// Copyright (c) 2022 Temporal Technologies Inc.  All rights reserved.
+//
+// Copyright (c) 2020 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package batcher
+
+import (
+	"context"
+	"errors"
+
+	"go.temporal.io/api/serviceerror"
+	"go.temporal.io/api/workflowservice/v1"
+	"go.temporal.io/sdk/activity"
+	sdkclient "go.temporal.io/sdk/client"
+	"golang.org/x/time/rate"
+
+	"go.temporal.io/server/common"
+	"go.temporal.io/server/common/dynamicconfig"
+	"go.temporal.io/server/common/log"
+	"go.temporal.io/server/common/log/tag"
+	"go.temporal.io/server/common/metrics"
+	"go.temporal.io/server/common/namespace"
+)
+
+var (
+	errNamespaceMismatch = errors.New("namespace mismatch")
+)
+
+type activities struct {
+	activityDeps
+	namespace   namespace.Name
+	namespaceID namespace.ID
+	rps         dynamicconfig.IntPropertyFnWithNamespaceFilter
+	concurrency dynamicconfig.IntPropertyFnWithNamespaceFilter
+}
+
+func (a *activities) checkNamespace(namespace string) error {
+	// Ignore system namespace for backward compatibility.
+	// TODO: Remove the system namespace special handling after 1.19+
+	if namespace != a.namespace.String() && a.namespace.String() != common.SystemLocalNamespace {
+		return errNamespaceMismatch
+	}
+	return nil
+}
+
+// BatchActivity is activity for processing batch operation
+func (a *activities) BatchActivity(ctx context.Context, batchParams BatchParams) (HeartBeatDetails, error) {
+	logger := a.getActivityLogger(ctx)
+	hbd := HeartBeatDetails{}
+	if err := a.checkNamespace(batchParams.Namespace); err != nil {
+		return hbd, err
+	}
+
+	sdkClient, err := a.ClientFactory.NewClient(batchParams.Namespace, logger)
```

Comment:
we don't want a long-lived client? is the idea that it will be idle most of the time?

Reply:
Just want to focus on enable per ns worker for batcher in this change. I can update batcher workflow logic in a different PR.